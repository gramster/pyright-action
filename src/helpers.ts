import * as core from '@actions/core';
import * as httpClient from '@actions/http-client';
import * as tc from '@actions/tool-cache';
import * as path from 'path';
import SemVer from 'semver/classes/semver';
import stringArgv from 'string-argv';

import { version as actionVersion } from '../package.json';
import { NpmRegistryResponse, parseNpmRegistryResponse } from './schema';

export function getActionVersion() {
    return actionVersion;
}

export function getNodeInfo() {
    return {
        version: process.version,
        execPath: process.execPath,
    };
}

export async function getArgs() {
    const pyrightInfo = await getPyrightInfo();
    const pyrightPath = await downloadPyright(pyrightInfo);

    const args = [path.join(pyrightPath, 'package', 'index.js')];

    const workingDirectory = core.getInput('working-directory');

    const noComments = getBooleanInput('no-comments', false);
    if (!noComments) {
        args.push('--outputjson');
    }

    const treatPartialAsWarning = getBooleanInput('warn-partial', false);
    const pythonPlatform = core.getInput('python-platform');
    if (pythonPlatform) {
        args.push('--pythonplatform');
        args.push(pythonPlatform);
    }

    const pythonVersion = core.getInput('python-version');
    if (pythonVersion) {
        args.push('--pythonversion');
        args.push(pythonVersion);
    }

    const typeshedPath = core.getInput('typeshed-path');
    if (typeshedPath) {
        args.push('--typeshed-path');
        args.push(typeshedPath);
    }

    const venvPath = core.getInput('venv-path');
    if (venvPath) {
        args.push('--venv-path');
        args.push(venvPath);
    }

    const project = core.getInput('project');
    if (project) {
        args.push('--project');
        args.push(project);
    }

    const lib = getBooleanInput('lib', false);
    if (lib) {
        args.push('--lib');
    }

    const warnings = getBooleanInput('warnings', false);
    if (warnings) {
        args.push('--warnings');
    }

    const verifyTypes = core.getInput('verify-types');
    if (verifyTypes) {
        args.push('--verifytypes');
        args.push(verifyTypes);
    }

    const extraArgs = core.getInput('extra-args');
    if (extraArgs) {
        args.push(...stringArgv(extraArgs));
    }

    return {
        workingDirectory,
        noComments,
        treatPartialAsWarning,
        pyrightVersion: pyrightInfo.version,
        args,
    };
}

function getBooleanInput(name: string, defaultValue: boolean): boolean {
    const input = core.getInput(name);
    if (!input) {
        return defaultValue;
    }
    return input.toUpperCase() === 'TRUE';
}

const pyrightToolName = 'pyright';

async function downloadPyright(info: NpmRegistryResponse): Promise<string> {
    // Note: this only works because the pyright package doesn't have any
    // dependencies. If this ever changes, we'll have to actually install it.
    const found = tc.find(pyrightToolName, info.version);
    if (found) {
        return found;
    }

    const tarballPath = await tc.downloadTool(info.dist.tarball);
    const extractedPath = await tc.extractTar(tarballPath);
    return await tc.cacheDir(extractedPath, pyrightToolName, info.version);
}

async function getPyrightInfo(): Promise<NpmRegistryResponse> {
    const version = getPyrightVersion();
    const client = new httpClient.HttpClient();
    const resp = await client.get(`https://registry.npmjs.org/pyright/${version}`);
    const body = await resp.readBody();
    if (resp.message.statusCode !== httpClient.HttpCodes.OK) {
        throw new Error(body);
    }
    return parseNpmRegistryResponse(JSON.parse(body));
}

function getPyrightVersion() {
    const versionSpec = core.getInput('version');
    if (versionSpec) {
        return new SemVer(versionSpec).format();
    }
    return 'latest';
}
