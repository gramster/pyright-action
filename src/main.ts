import * as core from '@actions/core';
import * as command from '@actions/core/lib/command';
import assert from 'assert';
import * as cp from 'child_process';

import { getActionVersion, getArgs, getNodeInfo } from './helpers';
import { Diagnostic, isEmptyRange, parseReport } from './schema';

export async function main() {
    try {
        const node = getNodeInfo();
        const { workingDirectory, noComments, treatPartialAsWarning, pyrightVersion, args } = await getArgs();
        if (workingDirectory) {
            process.chdir(workingDirectory);
        }

        core.info(`pyright ${pyrightVersion}, node ${node.version}, pyright-action ${getActionVersion()}`);
        core.info(`${node.execPath} ${args.join(' ')}`);

        // We check for --verifytypes as an arg instead of a flag because it may have
        // been passed via extra-args.
        if (noComments || args.includes('--verifytypes')) {
            // If comments are disabled, there's no point in directly processing the output,
            // as it's only used for comments.
            // If we're running the type verifier, there's no guarantee that we can even act
            // on the output besides the exit code.
            //
            // So, in either case, just directly run pyright and exit with its status.
            const { status } = cp.spawnSync(node.execPath, args, {
                stdio: ['ignore', 'inherit', 'inherit'],
            });

            if (status !== 0) {
                core.setFailed(`Exit code ${status!}`);
            }
            return;
        }

        let { status, stdout } = cp.spawnSync(process.execPath, args, {
            encoding: 'utf-8',
            stdio: ['ignore', 'pipe', 'inherit'],
        });

        if (!stdout.trim()) {
            // Process crashed. stderr was inherited, so just mark the step as failed.
            core.setFailed(`Exit code ${status!}`);
            return;
        }

        const report = parseReport(JSON.parse(stdout));
        let { errorCount, warningCount, informationCount } = report.summary;

        report.generalDiagnostics.forEach((diag) => {
            if (treatPartialAsWarning && diag.severity === 'error') {
                if (diag.message.includes('partially unknown')) {
                    diag.severity = 'warning';
                    errorCount -= 1;
                    warningCount += 1;
                    if (errorCount === 0) {
                        status = 0;
                    }
                }
            }

            core.info(diagnosticToString(diag, /* forCommand */ false));

            if (diag.severity === 'information') {
                return;
            }

            const line = diag.range?.start.line ?? 0;
            const col = diag.range?.start.character ?? 0;
            const message = diagnosticToString(diag, /* forCommand */ true);

            // This is technically a log line and duplicates the core.info above,
            // but we want to have the below look nice in commit comments.
            command.issueCommand(
                diag.severity,
                {
                    file: diag.file,
                    line: line + 1,
                    col: col + 1,
                },
                message
            );
        });

        core.info(
            [
                pluralize(errorCount, 'error', 'errors'),
                pluralize(warningCount, 'warning', 'warnings'),
                pluralize(informationCount, 'information', 'informations'),
            ].join(', ')
        );

        if (status !== 0) {
            core.setFailed(pluralize(errorCount, 'error', 'errors'));
        }
    } catch (e) {
        assert(typeof e === 'string' || e instanceof Error);
        core.setFailed(e);
    }
}

// Copied from pyright, with modifications.
function diagnosticToString(diag: Diagnostic, forCommand: boolean): string {
    let message = '';

    if (!forCommand) {
        if (diag.file) {
            message += `${diag.file}:`;
        }
        if (diag.range && !isEmptyRange(diag.range)) {
            message += `${diag.range.start.line + 1}:${diag.range.start.character + 1} -`;
        }
        message += ` ${diag.severity}: `;
    }

    message += diag.message;

    if (diag.rule) {
        message += ` (${diag.rule})`;
    }

    return message;
}

function pluralize(n: number, singular: string, plural: string) {
    return `${n} ${n === 1 ? singular : plural}`;
}
