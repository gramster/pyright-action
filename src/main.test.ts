import * as core from '@actions/core';
import * as command from '@actions/core/lib/command';
import * as cp from 'child_process';

import * as helpers from './helpers';

const nodeVersion = 'v16.14.2';
const nodeExecPath = '/path/to/node';
const pyrightVersion = '1.1.240';

jest.mock('@actions/core');
const mockedCore = jest.mocked(core);
jest.mock('@actions/core/lib/command');
const mockedCommand = jest.mocked(command);
jest.mock('child_process');
const mockedCp = jest.mocked(cp);
jest.mock('./helpers');
const mockedHelpers = jest.mocked(helpers);

const mockedProcessChdir = jest.spyOn(process, 'chdir');

import { main } from './main';
import { Report } from './schema';

beforeEach(() => {
    jest.clearAllMocks();
    mockedHelpers.getNodeInfo.mockReturnValue({
        version: nodeVersion,
        execPath: nodeExecPath,
    });
    mockedHelpers.getActionVersion.mockReturnValue('1.1.0');
    mockedProcessChdir.mockReturnValue(undefined); // This is a spy mock; prevent real process from being called.
});

afterEach(() => {
    expect(mockedProcessChdir.mock.calls).toMatchSnapshot('process.chdir');
    expect(mockedCore.setFailed.mock.calls).toMatchSnapshot('core.setFailed');
    expect(mockedCore.info.mock.calls).toMatchSnapshot('core.info');
    expect(mockedCommand.issueCommand.mock.calls).toMatchSnapshot('command.issueCommand');
    expect(mockedCp.spawnSync.mock.calls).toMatchSnapshot('cp.spawnSync');
});

test('thrown error at first call', async () => {
    mockedHelpers.getNodeInfo.mockImplementation(() => {
        throw new Error('oops');
    });

    await main();
});

describe('no comments', () => {
    const args = ['/path/to/pyright/dist/index.js', '--outputjson'];
    const wd = '/some/wd';

    beforeEach(() => {
        mockedHelpers.getArgs.mockResolvedValue({
            noComments: true,
            workingDirectory: wd,
            pyrightVersion,
            args,
        });
    });

    test('success', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: '',
            stderr: '',
            status: 0,
            signal: null,
        }));

        await main();
    });

    test('failure', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: '',
            stderr: '',
            status: 1,
            signal: null,
        }));

        await main();
    });
});

describe('with comments', () => {
    const args = ['/path/to/pyright/dist/index.js', '--outputjson'];

    beforeEach(() => {
        mockedHelpers.getArgs.mockResolvedValue({
            noComments: false,
            workingDirectory: '',
            pyrightVersion,
            args,
        });
    });

    test('failure', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: '',
            stderr: '',
            status: 2,
            signal: null,
        }));

        await main();
    });

    test('unparsable json', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: '}',
            stderr: '',
            status: 0,
            signal: null,
        }));

        await main();
    });

    test('invalid stdout', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: '{}',
            stderr: '',
            status: 0,
            signal: null,
        }));

        await main();
    });

    test('no diagnostics', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: reportToString({
                generalDiagnostics: [],
                summary: {
                    errorCount: 0,
                    warningCount: 0,
                    informationCount: 0,
                },
            }),
            stderr: '',
            status: 0,
            signal: null,
        }));

        await main();
    });

    test('with diagnostics', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: reportToString({
                generalDiagnostics: [
                    {
                        file: '/path/to/file1.py',
                        range: { start: { line: 0, character: 0 }, end: { line: 1, character: 1 } },
                        severity: 'error',
                        message: 'some error',
                    },
                    {
                        file: '/path/to/file2.py',
                        range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
                        severity: 'warning',
                        message: 'some warning',
                    },
                    {
                        file: '/path/to/file3.py',
                        range: { start: { line: 0, character: 0 }, end: { line: 1, character: 1 } },
                        severity: 'information',
                        message: 'some info',
                        rule: 'reportSomeInformation',
                    },
                    {
                        file: '/path/to/file3.py',
                        severity: 'warning',
                        message: 'another warning',
                        rule: 'reportSomeWarning',
                    },
                    {
                        file: '/path/to/file1.py',
                        range: { start: { line: 5, character: 9 }, end: { line: 5, character: 15 } },
                        severity: 'error',
                        message: 'some error',
                    },
                ],
                summary: {
                    errorCount: 2,
                    warningCount: 2,
                    informationCount: 1,
                },
            }),
            stderr: '',
            status: 1,
            signal: null,
        }));

        await main();
    });

    test('with 1 each', async () => {
        mockedCp.spawnSync.mockImplementation(() => ({
            pid: -1,
            output: [],
            stdout: reportToString({
                generalDiagnostics: [
                    {
                        file: '/path/to/file1.py',
                        range: { start: { line: 0, character: 0 }, end: { line: 1, character: 1 } },
                        severity: 'error',
                        message: 'some error',
                    },
                    {
                        file: '/path/to/file2.py',
                        range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
                        severity: 'warning',
                        message: 'some warning',
                    },
                    {
                        file: '/path/to/file3.py',
                        range: { start: { line: 0, character: 0 }, end: { line: 1, character: 1 } },
                        severity: 'information',
                        message: 'some info',
                        rule: 'reportSomeInformation',
                    },
                ],
                summary: {
                    errorCount: 1,
                    warningCount: 1,
                    informationCount: 1,
                },
            }),
            stderr: '',
            status: 1,
            signal: null,
        }));

        await main();
    });
});

function reportToString(report: Report): string {
    return JSON.stringify(report);
}
