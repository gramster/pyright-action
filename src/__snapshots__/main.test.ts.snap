// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no comments failure: command.issueCommand 1`] = `Array []`;

exports[`no comments failure: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
]
`;

exports[`no comments failure: core.setFailed 1`] = `
Array [
  Array [
    "Exit code 1",
  ],
]
`;

exports[`no comments failure: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "stdio": Array [
        "ignore",
        "inherit",
        "inherit",
      ],
    },
  ],
]
`;

exports[`no comments failure: process.chdir 1`] = `
Array [
  Array [
    "/some/wd",
  ],
]
`;

exports[`no comments success: command.issueCommand 1`] = `Array []`;

exports[`no comments success: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
]
`;

exports[`no comments success: core.setFailed 1`] = `Array []`;

exports[`no comments success: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "stdio": Array [
        "ignore",
        "inherit",
        "inherit",
      ],
    },
  ],
]
`;

exports[`no comments success: process.chdir 1`] = `
Array [
  Array [
    "/some/wd",
  ],
]
`;

exports[`thrown error at first call: command.issueCommand 1`] = `Array []`;

exports[`thrown error at first call: core.info 1`] = `Array []`;

exports[`thrown error at first call: core.setFailed 1`] = `
Array [
  Array [
    [Error: oops],
  ],
]
`;

exports[`thrown error at first call: cp.spawnSync 1`] = `Array []`;

exports[`thrown error at first call: process.chdir 1`] = `Array []`;

exports[`with comments failure: command.issueCommand 1`] = `Array []`;

exports[`with comments failure: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
]
`;

exports[`with comments failure: core.setFailed 1`] = `
Array [
  Array [
    "Exit code 2",
  ],
]
`;

exports[`with comments failure: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "encoding": "utf-8",
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`with comments failure: process.chdir 1`] = `Array []`;

exports[`with comments invalid stdout: command.issueCommand 1`] = `Array []`;

exports[`with comments invalid stdout: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
]
`;

exports[`with comments invalid stdout: core.setFailed 1`] = `
Array [
  Array [
    [ValitaError: missing_value at .summary (missing value) (+ 1 other issue)],
  ],
]
`;

exports[`with comments invalid stdout: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "encoding": "utf-8",
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`with comments invalid stdout: process.chdir 1`] = `Array []`;

exports[`with comments no diagnostics: command.issueCommand 1`] = `Array []`;

exports[`with comments no diagnostics: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
  Array [
    "0 errors, 0 warnings, 0 informations",
  ],
]
`;

exports[`with comments no diagnostics: core.setFailed 1`] = `Array []`;

exports[`with comments no diagnostics: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "encoding": "utf-8",
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`with comments no diagnostics: process.chdir 1`] = `Array []`;

exports[`with comments unparsable json: command.issueCommand 1`] = `Array []`;

exports[`with comments unparsable json: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
]
`;

exports[`with comments unparsable json: core.setFailed 1`] = `
Array [
  Array [
    [SyntaxError: Unexpected token } in JSON at position 0],
  ],
]
`;

exports[`with comments unparsable json: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "encoding": "utf-8",
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`with comments unparsable json: process.chdir 1`] = `Array []`;

exports[`with comments with 1 each: command.issueCommand 1`] = `
Array [
  Array [
    "error",
    Object {
      "col": 1,
      "file": "/path/to/file1.py",
      "line": 1,
    },
    "some error",
  ],
  Array [
    "warning",
    Object {
      "col": 1,
      "file": "/path/to/file2.py",
      "line": 1,
    },
    "some warning",
  ],
]
`;

exports[`with comments with 1 each: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
  Array [
    "/path/to/file1.py:1:1 - error: some error",
  ],
  Array [
    "/path/to/file2.py: warning: some warning",
  ],
  Array [
    "/path/to/file3.py:1:1 - information: some info (reportSomeInformation)",
  ],
  Array [
    "1 error, 1 warning, 1 information",
  ],
]
`;

exports[`with comments with 1 each: core.setFailed 1`] = `
Array [
  Array [
    "1 error",
  ],
]
`;

exports[`with comments with 1 each: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "encoding": "utf-8",
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`with comments with 1 each: process.chdir 1`] = `Array []`;

exports[`with comments with diagnostics: command.issueCommand 1`] = `
Array [
  Array [
    "error",
    Object {
      "col": 1,
      "file": "/path/to/file1.py",
      "line": 1,
    },
    "some error",
  ],
  Array [
    "warning",
    Object {
      "col": 1,
      "file": "/path/to/file2.py",
      "line": 1,
    },
    "some warning",
  ],
  Array [
    "warning",
    Object {
      "col": 1,
      "file": "/path/to/file3.py",
      "line": 1,
    },
    "another warning (reportSomeWarning)",
  ],
  Array [
    "error",
    Object {
      "col": 10,
      "file": "/path/to/file1.py",
      "line": 6,
    },
    "some error",
  ],
]
`;

exports[`with comments with diagnostics: core.info 1`] = `
Array [
  Array [
    "pyright 1.1.240, node v16.14.2, pyright-action 1.1.0",
  ],
  Array [
    "/path/to/node /path/to/pyright/dist/index.js --outputjson",
  ],
  Array [
    "/path/to/file1.py:1:1 - error: some error",
  ],
  Array [
    "/path/to/file2.py: warning: some warning",
  ],
  Array [
    "/path/to/file3.py:1:1 - information: some info (reportSomeInformation)",
  ],
  Array [
    "/path/to/file3.py: warning: another warning (reportSomeWarning)",
  ],
  Array [
    "/path/to/file1.py:6:10 - error: some error",
  ],
  Array [
    "2 errors, 2 warnings, 1 information",
  ],
]
`;

exports[`with comments with diagnostics: core.setFailed 1`] = `
Array [
  Array [
    "2 errors",
  ],
]
`;

exports[`with comments with diagnostics: cp.spawnSync 1`] = `
Array [
  Array [
    "/path/to/node",
    Array [
      "/path/to/pyright/dist/index.js",
      "--outputjson",
    ],
    Object {
      "encoding": "utf-8",
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`with comments with diagnostics: process.chdir 1`] = `Array []`;
